libraries/script-engine/src/ScriptEngine.cpp

```diff
@@ -135,38 +135,53 @@ QString encodeEntityIdIntoEntityUrl(const QString& url, const QString& entityID)
     return url + " [EntityID:" + entityID + "]";
 }
 
-static bool hasCorrectSyntax(const QScriptProgram& program, ScriptEngine* reportingEngine) {
-    const auto syntaxCheck = QScriptEngine::checkSyntax(program.sourceCode());
-    if (syntaxCheck.state() != QScriptSyntaxCheckResult::Valid) {
+QScriptProgram ScriptEngine::compileScript(const QString& sourceCode, const QString& fileName, const int lineNumber) {
+    QScriptProgram program(sourceCode, fileName, lineNumber);
+    const auto syntaxCheck = checkSyntax(program.sourceCode());
+    if (syntaxCheck.state() != syntaxCheck.Valid) {
         const auto error = syntaxCheck.errorMessage();
         const auto line = QString::number(syntaxCheck.errorLineNumber());
         const auto column = QString::number(syntaxCheck.errorColumnNumber());
         const auto message = QString("[SyntaxError] %1 in %2:%3(%4)").arg(error, program.fileName(), line, column);
-        reportingEngine->scriptErrorMessage(qPrintable(message));
-        return false;
+        scriptErrorMessage(qPrintable(message));
+        return QScriptProgram();
     }
-    return true;
+    return program;
 }
 
-static bool hadUncaughtExceptions(QScriptEngine& engine, const QString& fileName, ScriptEngine* reportingEngine, QString* exceptionMessage = nullptr) {
-    if (engine.hasUncaughtException()) {
-        const auto backtrace = engine.uncaughtExceptionBacktrace();
-        const auto exception = engine.uncaughtException().toString();
-        const auto line = QString::number(engine.uncaughtExceptionLineNumber());
-        engine.clearExceptions();
+QString ScriptEngine::reportUncaughtExceptions(const QString& fileName, QScriptEngine* engine) {
+    if (engine == nullptr) {
+        engine = this;
+    }
+    if (hasUncaughtException()) {
+        const auto backtrace = engine->uncaughtExceptionBacktrace();
+        const auto exception = engine->uncaughtException().toString();
+        const auto line = QString::number(engine->uncaughtExceptionLineNumber());
+        const auto nestedEval = engine == this && evaluatePending();
+        if (!nestedEval) {
+            // consume the exception only if the top-level evaluate/evaluateInClosure
+            engine->clearExceptions();
+        }
```

Comment:
Yep, or at least that was the theory.

From what I can gather, exceptions were being cleared prematurely as a workaround to prevent the same error messages from getting reported more than once (like as the call stack unwinds from nested includes).

The Qt-recommended way of detecting uncaught exceptions is http://doc.qt.io/qt-5.6/qscriptengineagent.html#exceptionThrow, which I was tempted to do here (since letting exceptions through is critical to `require` behavior), but this change set is already pretty bulky.